docs/architecture/adr-009-evidence-module.md

```diff
@@ -0,0 +1,179 @@
+# ADR 009: Evidence Module
+
+## Changelog
+
+- 2019 July 31: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+In order to support building highly secure, robust and interoperable blockchain
+applications, it is vital for the Cosmos SDK to expose a mechanism in which arbitrary
+evidence can be submitted, evaluated and verified resulting in some agreed upon
+penalty for any misbehavior committed by a validator, such as equivocation (double-voting),
+signing when unbonded, signing an incorrect state transition (in the future), etc.
+Furthermore, such a mechanism is paramount for any
+[IBC](https://github.com/cosmos/ics/blob/master/ibc/1_IBC_ARCHITECTURE.md) or
+cross-chain validation protocol implementation in order to support the ability
+for any misbehavior to be relayed back from a collateralized chain to a primary
+chain so that the equivocating validator(s) can be slashed.
+
+## Decision
+
+We will implement an evidence module in the Cosmos SDK supporting the following
+functionality:
+
+- Provide developers with the abstractions and interfaces necessary to define
+custom evidence messages, message handlers, and methods to slash and penalize
+accordingly for misbehavior.
+- Support the ability to route evidence messages to handlers in any module to
+determine the validity of submitted misbehavior.
+- Support the ability, through governance, to modify slashing penalties of any
+evidence type.
+- Querier implementation to support querying params, evidence types, params, and
+all submitted valid misbehavior.
+
+### Types
+
+First, we define the `Evidence` interface type. The `x/evidence` module may implement
+its own types that can be used by many chains (e.g. `CounterFactualEvidence`).
+In addition, other modules may implement their own `Evidence` types in a similar
+manner in which governance is extensible. It is important to note any concrete
+type implementing the `Evidence` interface may include arbitrary fields such as
+an infraction time. We want the `Evidence` type to remain as flexible as possible.
+
+When submitting evidence to the `x/evidence` module, the concrete type must provide
+the validator's consensus address, which should be known by the `x/slashing`
+module (assuming the infraction is valid), the height at which the infraction
+occurred and the validator's power at same height in which the infraction occurred.
+
+```go
+type Evidence interface {
+  Route() string
+  Type() string
+  String() string
+  ValidateBasic() Error
+
+  // The consensus address of the malicious validator at time of infraction
+  GetConsensusAddress() ConsAddress
```

Comment:
https://golang.org/doc/effective_go.html#Getters

Reply:
Yes, I know. But typically we have field members as public due to amino encoding requirements and we want to also establish a contract of what evidence "must" provide.